src/librustc_infer/traits/wf.rs

```diff
@@ -319,16 +319,17 @@ impl<'a, 'tcx> WfPredicates<'a, 'tcx> {
             let trait_assoc_items: Vec<_> =
                 tcx.associated_items(trait_ref.def_id).in_definition_order().copied().collect();
 
-            let predicates = obligations.iter().map(|obligation| obligation.predicate).collect();
-            let implied_obligations = traits::elaborate_predicates(tcx, predicates);
-            let implied_obligations = implied_obligations.map(|pred| {
+            let implied_obligations = traits::elaborate_obligations(tcx, obligations.clone());
+            let implied_obligations = implied_obligations.map(|obligation| {
                 let mut cause = cause.clone();
+                cause.code =
+                    traits::ObligationCauseCode::DerivedCauseCode(Box::new(obligation.cause.code));
```

Comment:
So this seems to be the only place where `DerivedCauseCode` is created?

I think most of the test output changes are because of the extra `Span` in `ItemObligation`, *not* this, am I right?

If so, it might be worth splitting this PR into the non-`DerivedCauseCode` changes and the `DerivedCauseCode` changes, because the former seem far less controversial (and easier to understand) to me.

Reply:
"most" might be the case. Removing this part will greatly diminish the benefit of this change, but I can oblige and let the PRs evolve independently :)